### PR TITLE
[RFC][NO MERGE] flatten block nodes 

### PR DESCRIPTION
**Pull request by potatowagon**
draft only

see issue #4465

test case:
```
<div>
  a
  <div>
    b
    <div>
      c
      <div>
        <div>
        </div>
        z
      </div>
    </div>
    d
    e
  </div>
  fg
</div>
```

https://github.com/facebook/lexical/assets/19604232/29addc46-ce88-4477-b8e7-30209ad481f5

---
This will break nested block nodes like list nodes, because we are flattening them

fixed in commit a1f17f6db537f24146fbd53bb4eaf9ef72158fed which only limits the unnesting to block nodes in para nodes

test case
```
<ol class="PlaygroundEditorTheme__ol1"><li value="1" class="PlaygroundEditorTheme__listItem"><span style="white-space: pre-wrap;">ssvsdvs</span></li><li value="2" class="PlaygroundEditorTheme__listItem"><span style="white-space: pre-wrap;">vsvsvsv</span></li><li value="3" class="PlaygroundEditorTheme__listItem"><span style="white-space: pre-wrap;">svsvsvs</span></li><li value="4" class="PlaygroundEditorTheme__listItem PlaygroundEditorTheme__nestedListItem"><ol class="PlaygroundEditorTheme__ol2"><li value="1" class="PlaygroundEditorTheme__listItem"><span style="white-space: pre-wrap;">fvsdvsdv</span></li><li value="2" class="PlaygroundEditorTheme__listItem"><span style="white-space: pre-wrap;">sdvsvdsv</span></li><li value="3" class="PlaygroundEditorTheme__listItem"><span style="white-space: pre-wrap;">sdvsdvsd</span></li></ol></li><li value="4" class="PlaygroundEditorTheme__listItem"><span style="white-space: pre-wrap;">sdvsdvdsv</span></li><li value="5" class="PlaygroundEditorTheme__listItem"><span style="white-space: pre-wrap;">sdvsdvdsv</span></li><li value="6" class="PlaygroundEditorTheme__listItem"><span style="white-space: pre-wrap;">sdvsdvsdv</span></li></ol>
```
in codepen:

Screenshot 2024-04-05 at 4 38 29 PM

in lexical:

Screenshot 2024-04-05 at 4 38 47 PM

after a1f17f6db537f24146fbd53bb4eaf9ef72158fed fix 

Screenshot 2024-04-05 at 6 08 48 PM

---
table in a div

test case:
```
<div>
<table class="PlaygroundEditorTheme__table"><colgroup><col><col></colgroup><tbody><tr><th class="PlaygroundEditorTheme__tableCell PlaygroundEditorTheme__tableCellHeader" style="border: 1px solid black; width: 350px; vertical-align: top; text-align: start; background-color: rgb(242, 243, 245);"><p class="PlaygroundEditorTheme__paragraph"><br></p></th><th class="PlaygroundEditorTheme__tableCell PlaygroundEditorTheme__tableCellHeader" style="width: 398px; border: 1px solid black; vertical-align: top; text-align: start; background-color: rgb(242, 243, 245);"><p class="PlaygroundEditorTheme__paragraph"><br></p></th></tr><tr><th class="PlaygroundEditorTheme__tableCell PlaygroundEditorTheme__tableCellHeader" style="border: 1px solid black; width: 350px; vertical-align: top; text-align: start; background-color: rgb(242, 243, 245);"><p class="PlaygroundEditorTheme__paragraph"><br></p></th><td class="PlaygroundEditorTheme__tableCell" style="width: 398px; border: 1px solid black; vertical-align: top; text-align: start;"><ol class="PlaygroundEditorTheme__ol1"><li value="1" class="PlaygroundEditorTheme__listItem"><span style="white-space: pre-wrap;">sdcsdcsccsc</span></li><li value="2" class="PlaygroundEditorTheme__listItem"><span style="white-space: pre-wrap;">dscsc</span></li><li value="3" class="PlaygroundEditorTheme__listItem"><span style="white-space: pre-wrap;">sdcsc</span></li><li value="4" class="PlaygroundEditorTheme__listItem"><span style="white-space: pre-wrap;">scscs</span></li><li value="5" class="PlaygroundEditorTheme__listItem"><span style="white-space: pre-wrap;">csdcs</span></li><li value="6" class="PlaygroundEditorTheme__listItem"><span style="white-space: pre-wrap;">dcscsc</span></li><li value="7" class="PlaygroundEditorTheme__listItem"><span style="white-space: pre-wrap;">vsdvsdv</span></li><li value="8" class="PlaygroundEditorTheme__listItem"><span style="white-space: pre-wrap;">sdsv</span></li><li value="9" class="PlaygroundEditorTheme__listItem"><span style="white-space: pre-wrap;">sdvsvsdv</span></li><li value="10" class="PlaygroundEditorTheme__listItem"><span style="white-space: pre-wrap;">svsdvds</span></li><li value="11" class="PlaygroundEditorTheme__listItem PlaygroundEditorTheme__nestedListItem"><ol class="PlaygroundEditorTheme__ol2"><li value="1" class="PlaygroundEditorTheme__listItem"><span style="white-space: pre-wrap;">svsdvsd</span></li><li value="2" class="PlaygroundEditorTheme__listItem"><span style="white-space: pre-wrap;">vsvdsv</span></li><li value="3" class="PlaygroundEditorTheme__listItem PlaygroundEditorTheme__nestedListItem"><ol class="PlaygroundEditorTheme__ol3"><li value="1" class="PlaygroundEditorTheme__listItem"><span style="white-space: pre-wrap;">dfdbfdb</span></li><li value="2" class="PlaygroundEditorTheme__listItem"><span style="white-space: pre-wrap;">dfbdfbdfb</span></li><li value="3" class="PlaygroundEditorTheme__listItem"><span style="white-space: pre-wrap;">f</span><br><span style="white-space: pre-wrap;">dfbdf</span><br><span style="white-space: pre-wrap;">dfbdfb</span><br><span style="white-space: pre-wrap;">dfbdfb</span></li><li value="4" class="PlaygroundEditorTheme__listItem"></li></ol></li></ol></li></ol><p class="PlaygroundEditorTheme__paragraph" dir="ltr"><br></p><p class="PlaygroundEditorTheme__paragraph"><br></p><p class="PlaygroundEditorTheme__paragraph"><br></p></td></tr></tbody></table>
</div>
```
codepen:
Screenshot 2024-04-05 at 5 57 40 PM

lexical:

Screenshot 2024-04-05 at 5 58 28 PM

---
handle divs in li

```
<ol>
  <li>
    <div>
      1
      <div>
        2
        <div>
          <div>
            <div>
              3
              <div>
  </li>
  <div>
    <li>
      <div>
        1
        <div>
          2
          <div>
            <div>
              <div>
                3
                <div>
    </li>
  </div>
</ol>
```

### CODE REVIEW

File: packages/lexical/src/nodes/LexicalParagraphNode.ts [hunk 1]
```diff
@@ -60,6 +60,12 @@ export class ParagraphNode extends ElementNode {
 
   static importDOM(): DOMConversionMap | null {
     return {
+      div: (node: Node) => ({
+        conversion: (e) => {
+          return convertParagraphElement(e);
+        },
+        priority: 4,
+      }),
```
[the remark this answers is not in the document]
Comment (potatowagon) [reply packages/lexical/src/nodes/LexicalParagraphNode.ts:68]: currently these other nodes have conversion fns for div:

1. CodeNode.ts:146
2. EquationNode.tsx:110
3.  TweetNode.tsx:161
4. CollapsibleContentNode.ts:51
5.  LayoutContainerNode.ts:85

[the remark this answers is not in the document]
Comment (potatowagon) [reply packages/lexical/src/nodes/LexicalParagraphNode.ts:68]: for poc adding a div -> para conversion to all other node conversions as fallback

File: packages/lexical-html/src/index.ts [hunk 2]
```diff
@@ -168,6 +177,11 @@ function getConversionFunction(
         (currentConversion === null ||
           (currentConversion.priority || 0) < (domConversion.priority || 0))
       ) {
+        // [issue?] this means only one conversion function gets to run.
+        // If there are multiple nodes initialised with mapping to same html node, only the highest pri function gets ran.
```
Comment (acywatson) [inline packages/lexical-html/src/index.ts:181]: This is not correct (or if it is, there's a bug). The intent is that conversion functions run for a given HTML Element type and they decide if they can handle that element type or not. If they can, they return a node. If they can't they return null or something (can't recall the exact API) and then the next conversion for that element type runs.

Comment (potatowagon) [reply packages/lexical-html/src/index.ts:181]: > If they can, they return a node. If they can't they return null or something (can't recall the exact API) and then the next conversion for that element type runs.

ah got it! i prolly missed this part. lemme do some testing to confirm this behavior.

**Comment by potatowagon** [comment]
<ol>
  <li>
    1
    <div>
      2
    </div>
    3
  </li>
    <li>
    A
    <div>
      B
    </div>
    C
  </li>
</ol>

**Pull request closed without merging**